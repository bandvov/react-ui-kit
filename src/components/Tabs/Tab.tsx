import React from "react";
import Button from "../Button/Button";

export default function Tab({
  label,
  onClick,
  value,
  buttonStyles = {},
  activeButtonStyles = {},
  ...rest
}: {
  buttonStyles?: { [key: string]: string };
  activeButtonStyles?: { [key: string]: string };
  fullWidth?: boolean;
  label: string;
  onClick: (value: number | string) => void;
  value: string | number;
}) {
  return (
    <Button
      padding="1rem"
      style={value === label ? activeButtonStyles : buttonStyles}
      onClick={() => onClick(value)}
      color={value === label ? "" : "lightgray"}
      borderRadius="0"
      {...rest}
    >
      {label}
    </Button>
  );
}
