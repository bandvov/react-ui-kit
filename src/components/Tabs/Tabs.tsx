import React, { Dispatch, SetStateAction, useRef, useState } from "react";
import styled from "styled-components";
import Tab from "./Tab";

const StyledTabs = styled.div`
  display: flex;
  width: 50vw;
  position: relative;
  overflow-x: hidden;
  ::-webkit-scrollbar {
    width: 0;
  }
`;
export default function Tabs({
  value,
  onChange,
}: {
  value: string | number;
  onChange: Dispatch<SetStateAction<string>>;
}) {
  const containerRef = useRef<HTMLDivElement>(null);
  const childrenRef = useRef<HTMLDivElement>(null);
  const startXRef = useRef(0);

  const [offsetX, setOffsetX] = useState(0);

  const onTouchStart = (e: any) => {
    const clientX = e.clientX ? e.clientX : e.touches[0].clientX;

    startXRef.current = offsetX + clientX;

    window.addEventListener("mouseup", onTouchEnd);
    window.addEventListener("touchend", onTouchEnd);
    window.addEventListener("touchmove", onTouchMove);
    window.addEventListener("mousemove", onTouchMove);
  };

  const onTouchMove = (e: any) => {
    const currentX = e.clientX ? e.clientX : e.touches[0].clientX;

    const diff = startXRef.current - currentX;

    const currentOffsetLeft = childrenRef.current?.offsetLeft || 0;

    let newOffset = currentOffsetLeft + diff;
    if (newOffset < 0) {
      newOffset = 0;
    }
    if (
      newOffset >
      childrenRef.current!.offsetWidth - containerRef.current!.offsetWidth
    ) {
      newOffset =
        childrenRef.current!.offsetWidth - containerRef.current!.offsetWidth;
    }
    setOffsetX(newOffset);
  };

  const onTouchEnd = () => {
    window.removeEventListener("touchmove", onTouchMove);
    window.removeEventListener("mousemove", onTouchMove);
    window.removeEventListener("mouseup", onTouchEnd);
    window.removeEventListener("touchend", onTouchEnd);
  };

  return (
    <div style={{ display: "flex", flexDirection: "column" }}>
      <div>
        <div>{startXRef.current}</div>
        <div>{offsetX}</div>
      </div>
      <StyledTabs
        ref={containerRef}
        onTouchStart={onTouchStart}
        onMouseDown={onTouchStart}
        // onMouseUp={(e) => {}}
        // onTouchMove={(e: React.TouchEvent<HTMLDivElement>) => {}}
        style={{ border: "3px solix" }}
        onMouseLeave={onTouchEnd}
      >
        <div
          ref={childrenRef}
          style={{
            display: "flex",
            transform: `translateX(-${offsetX}px)`,
          }}
        >
          <Tab
            fullWidth
            value={value}
            variant="error"
            onClick={() => {
              console.log("hello");
            }}
            label="One"
          />
          <Tab
            value={value}
            fullWidth
            variant="success"
            onClick={(value: any) => alert("hello")}
            label="Two"
          />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab value={value} fullWidth onClick={() => {}} label="Three" />
          <Tab
            value={value}
            fullWidth
            variant="danger"
            onClick={() => {}}
            label="Three"
          />
        </div>
      </StyledTabs>
    </div>
  );
}
