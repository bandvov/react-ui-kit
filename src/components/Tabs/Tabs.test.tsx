import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Tabs from "./Tabs";

const tabItems = ["One", "Two", "Three"];

describe("Tabs", () => {
  test("Tabs should be in the document", () => {
    const handler = jest.fn();
    render(<Tabs items={tabItems} value={1} onChange={handler} />);
    const tabsContainer = screen.getByTestId("tabs");
    const tabs = screen.getAllByTestId("button");
    expect(tabsContainer).toMatchSnapshot();
    expect(tabs).toHaveLength(3);
  });
  test("Tab handler", () => {
    const handler = jest.fn();
    render(<Tabs items={tabItems} value={1} onChange={handler} />);
    const tabs = screen.getAllByTestId("button");
    fireEvent.click(tabs[0]);
    expect(handler).toHaveBeenCalledTimes(1);
    expect(handler).toHaveBeenCalledWith(0);
  });
});
