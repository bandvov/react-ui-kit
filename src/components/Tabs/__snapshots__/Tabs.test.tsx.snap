// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Tabs should be in the document 1`] = `
<div
  class="sc-gsDKAQ eJRZPP"
  data-testid="tabs"
>
  <button
    class="sc-bdvvtL feyXfz"
    color="lightgray"
    data-testid="button"
    type="button"
  >
    One
  </button>
  <button
    class="sc-bdvvtL ecqrqj"
    color=""
    data-testid="button"
    type="button"
  >
    Two
  </button>
  <button
    class="sc-bdvvtL feyXfz"
    color="lightgray"
    data-testid="button"
    type="button"
  >
    Three
  </button>
</div>
`;
